### PR TITLE
[1.2.0] add "RemoveEntry" and "ContainsEntry" APIs to MultiMap [API-478]

### PR DESCRIPTION
**Pull request by utku-caglayan**
add "RemoveEntry" and "ContainsEntry" APIs to MultiMap [API-478]

[API-478]: https://hazelcast.atlassian.net/browse/API-478?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by yuce — APPROVED** [review]
Looks good!

File: proxy_multi_map.go [hunk 1]
```diff
@@ -159,6 +159,25 @@ func (m *MultiMap) ContainsValue(ctx context.Context, value interface{}) (bool,
...
+	valueData, err := m.validateAndSerialize(value)
+	if err != nil {
+		return false, err
+	}
+	request := codec.EncodeMultiMapContainsEntryRequest(m.name, keyData, valueData, lid)
+	response, err := m.invokeOnKey(ctx, request, keyData)
+	if err != nil {
+		return false, err
+	}
+	return codec.DecodeMultiMapContainsKeyResponse(response), nil
```
Comment (srknzl) [inline proxy_multi_map.go:178]:
```suggestion
	return codec.DecodeMultiMapContainsEntryResponse(response), nil
```

File: multi_map_it_test.go [hunk 2]
```diff
@@ -124,6 +124,31 @@ func TestMultiMap_Remove(t *testing.T) {
...
 
+func TestMultiMap_RemoveEntry(t *testing.T) {
+	it.MultiMapTester(t, func(t *testing.T, m *hz.MultiMap) {
+		ctx := context.Background()
+		targetValue := "value"
+		otherValue := "other value"
+		it.Must(m.PutAll(ctx, "key", targetValue, otherValue))
+		value := it.MustValue(m.Get(ctx, "key"))
+		assert.ElementsMatch(t, value, []interface{}{targetValue, otherValue})
+		// Remove only one of the values that corresponds to the key
```
Comment (nevzatseferoglu) [inline multi_map_it_test.go:135]:
```suggestion
// Remove only one of the value that corresponds to the key.
```

File: multi_map_it_test.go [hunk 3]
```diff
@@ -124,6 +124,31 @@ func TestMultiMap_Remove(t *testing.T) {
...
+		ok, err := m.RemoveEntry(ctx, "key", targetValue)
+		if err != nil {
+			t.Fatal(err)
+		}
+		assert.True(t, ok)
+		remaining := it.MustValue(m.Get(ctx, "key"))
+		assert.Equal(t, []interface{}{otherValue}, remaining)
+		// Call should have no effect, expect "false" return value
```
Comment (nevzatseferoglu) [inline multi_map_it_test.go:143]:
```suggestion
// Call should have no effect, expects "false" as a return value.
```

File: proxy_multi_map.go [hunk 4]
```diff
@@ -159,6 +159,25 @@ func (m *MultiMap) ContainsValue(ctx context.Context, value interface{}) (bool,
 	}
 }
 
+// ContainsEntry returns true if the map contains an entry with the given key and value.
```
Comment (nevzatseferoglu) [inline proxy_multi_map.go:162]:
```suggestion
// ContainsEntry returns true if the MultiMap contains an entry with the given key and value.
```

File: proxy_multi_map.go [hunk 5]
```diff
@@ -159,6 +159,25 @@ func (m *MultiMap) ContainsValue(ctx context.Context, value interface{}) (bool,
...
+	response, err := m.invokeOnKey(ctx, request, keyData)
+	if err != nil {
+		return false, err
+	}
+	return codec.DecodeMultiMapContainsKeyResponse(response), nil
+}
+
 // Delete removes the mapping for a key from this multi-map if it is present.
 // Unlike remove(object), this operation does not return the removed value, which avoids the serialization cost of
 // the returned value. If the removed value will not be used, a delete operation is preferred over a remove
```
Comment (nevzatseferoglu) [inline proxy_multi_map.go:183]:
```suggestion
// the returned value. If the removed value will not be used, a delete operation is preferred over a remove.
```

**Review by nevzatseferoglu — CHANGES_REQUESTED** [review]
Looks good!

File: examples/multimap/main.go [hunk 6]
```diff
@@ -40,14 +39,18 @@ func main() {
...
 	// An interface slice contains the values.
 	// []interface {}{"value2", "value1"}
 	fmt.Printf("%#v", values)
+	// Check existence of an entry
+	ok, err := m.ContainsEntry(ctx, "key", "value1")
+	if err != nil {
+		log.Fatal(err)
+	}
+	fmt.Printf("entry found:%t", ok)
 }
```
[the remark this answers is not in the document]
Comment (utku-caglayan) [reply examples/multimap/main.go:56]: Thanks @srknzl for pointing this out. They will be addressed in #722 and #723 🙏

File: proxy_multi_map.go [hunk 7]
```diff
@@ -323,6 +342,25 @@ func (m *MultiMap) Remove(ctx context.Context, key interface{}) ([]interface{},
 	}
 }
 
+// RemoveEntry removes the specified value for the given key and returns true if call had an effect
```
[the remark this answers is not in the document]
Comment (utku-caglayan) [reply proxy_multi_map.go]: Fixed as suggested https://github.com/hazelcast/hazelcast-go-client/pull/725/commits/b58fa1b6acabcb86878ab48d0b9fa24b5c27d21c

File: proxy_multi_map.go (hunk 1 above)
Comment (utku-caglayan) [reply proxy_multi_map.go]: Great catch, thanks a lot. Fixed at https://github.com/hazelcast/hazelcast-go-client/pull/725/commits/b58fa1b6acabcb86878ab48d0b9fa24b5c27d21c

File: multi_map_it_test.go (hunk 2 above)
Comment (utku-caglayan) [reply multi_map_it_test.go]: Fixed at https://github.com/hazelcast/hazelcast-go-client/pull/725/commits/b58fa1b6acabcb86878ab48d0b9fa24b5c27d21c.

File: multi_map_it_test.go (hunk 3 above)
Comment (utku-caglayan) [reply multi_map_it_test.go]: Thanks, fixed to something similar https://github.com/hazelcast/hazelcast-go-client/pull/725/commits/b58fa1b6acabcb86878ab48d0b9fa24b5c27d21c

File: proxy_multi_map.go (hunk 4 above)
Comment (utku-caglayan) [reply proxy_multi_map.go]: Fixed at https://github.com/hazelcast/hazelcast-go-client/pull/725/commits/b58fa1b6acabcb86878ab48d0b9fa24b5c27d21c.

File: proxy_multi_map.go (hunk 5 above)
Comment (utku-caglayan) [reply proxy_multi_map.go]: Thanks! Fixed at https://github.com/hazelcast/hazelcast-go-client/pull/725/commits/b58fa1b6acabcb86878ab48d0b9fa24b5c27d21c.

**Review by srknzl — APPROVED** [review]
Thanks!

**Review by nevzatseferoglu — APPROVED** [review]

**Pull request merged**